File: tests/python/test_vtu.py
import pathlib
import tempfile

import numpy as np
import pytest

import meshioplusplus
from meshioplusplus.vtu import _vtu

from . import helpers

test_set = [
    # helpers.empty_mesh,
    helpers.line_mesh,
    helpers.tri_mesh,
    helpers.tri_mesh_one_cell,
    helpers.triangle6_mesh,
    helpers.quad_mesh,
    helpers.quad8_mesh,
    helpers.tri_quad_mesh,
    helpers.polygon_mesh,
    helpers.polygon_mesh_one_cell,
    helpers.polygon2_mesh,
    helpers.tet_mesh,
    helpers.tet10_mesh,
    helpers.hex_mesh,
    helpers.hex20_mesh,
    helpers.pyramid_mesh,
    helpers.wedge_mesh,
    helpers.polyhedron_mesh,
    helpers.lagrange_high_order_mesh,
    helpers.add_point_data(helpers.tri_mesh, 1),
    helpers.add_point_data(helpers.tri_mesh, 2),
    helpers.add_point_data(helpers.tri_mesh, 3),
    helpers.add_cell_data(helpers.tri_mesh, [("a", (), np.float64)]),
    helpers.add_cell_data(helpers.tri_quad_mesh, [("a", (), np.float64)]),
    helpers.add_cell_data(helpers.tri_mesh, [("a", (2,), np.float32)]),
    helpers.add_cell_data(helpers.tri_mesh, [("b", (3,), np.float64)]),
    helpers.add_cell_data(helpers.polygon_mesh, [("a", (), np.float32)]),
    helpers.add_cell_data(helpers.polyhedron_mesh, [("a", (2,), np.float32)]),
]


@pytest.mark.parametrize("mesh", test_set)
@pytest.mark.parametrize(
    "data_type", [(False, None), (True, None), (True, "lzma"), (True, "zlib")]
)
def test(mesh, data_type, tmp_path):
    binary, compression = data_type

    def writer(*args, **kwargs):
        return meshioplusplus.vtu.write(
            *args, binary=binary, compression=compression, **kwargs
        )

    # ASCII files are only meant for debugging, VTK stores only 11 digits
    # <https://gitlab.kitware.com/vtk/vtk/-/issues/17038#note_264052>
    tol = 1.0e-15 if binary else 1.0e-10
    helpers.write_read(tmp_path, writer, meshioplusplus.vtu.read, mesh, tol)


def test_generic_io(tmp_path):
    helpers.generic_io(tmp_path / "test.vtu")
    # With additional, insignificant suffix:
    helpers.generic_io(tmp_path / "test.0.vtu")


@pytest.mark.parametrize(
    "filename, ref_cells, ref_num_cells, ref_num_pnt",
    [
        ("00_raw_binary.vtu", "tetra", 162, 64),
        ("01_raw_binary_int64.vtu", "tetra", 162, 64),
        ("02_raw_compressed.vtu", "tetra", 162, 64),
    ],
)
def test_read_from_file(filename, ref_cells, ref_num_cells, ref_num_pnt):
    this_dir = pathlib.Path(__file__).resolve().parent
    filename = this_dir / "meshes" / "vtu" / filename

    mesh = meshioplusplus.read(filename)
    assert len(mesh.cells) == 1
    assert ref_cells == mesh.cells[0].type
    assert len(mesh.cells[0].data) == ref_num_cells
    assert len(mesh.points) == ref_num_pnt


# --- malformed-input / error-path coverage ---


def test_vtu_wrong_root_tag_raises(tmp_path):
    # A well-formed XML document that is not a VTKFile is rejected on both the
    # C++ and Python paths, so the ReadError surfaces through the shim.
    p = tmp_path / "bad.vtu"
    p.write_text('<?xml version="1.0"?><NotVTK></NotVTK>')
    with pytest.raises(meshioplusplus.ReadError):
        meshioplusplus.read(p, file_format="vtu")


def test_vtu_not_xml_raises(tmp_path):
    p = tmp_path / "bad.vtu"
    p.write_text("this is not xml at all")
    with pytest.raises(meshioplusplus.ReadError):
        meshioplusplus.read(p, file_format="vtu")


def test_polyhedra_mix_with_other_cell_types_both_engines():
    """VTU's `faceoffsets` carries -1 for a non-polyhedral cell, which is
    exactly how the format expresses a mixed mesh -- and an OpenFOAM mesh always
    mixes. The C++ and Python paths must agree, because Windows CI builds native
    paths off and runs the reference implementation.
    """
    pts = np.array(
        [
            [0, 0, 0],
            [1, 0, 0],
            [1, 1, 0],
            [0, 1, 0],
            [0, 0, 1],
            [1, 0, 1],
            [1, 1, 1],
            [0, 1, 1],
        ],
        float,
    )
    tet_faces = [
        np.array([0, 2, 1]),
        np.array([0, 1, 3]),
        np.array([1, 2, 3]),
        np.array([2, 0, 3]),
    ]
    mesh = meshioplusplus.Mesh(
        pts,
        [
            ("triangle", np.array([[4, 5, 6]])),
            ("polyhedron4", [tet_faces]),
        ],
    )

    for writer, reader, tag in (
        (_vtu.write, _vtu.read, "python"),
        (meshioplusplus.vtu.write, meshioplusplus.vtu.read, "shim"),
    ):
        with tempfile.TemporaryDirectory() as d:
            p = pathlib.Path(d) / "mixed.vtu"
            writer(p, mesh)
            out = reader(p)
            types = [c.type for c in out.cells]
            assert any(t.startswith("polyhedron") for t in types), (tag, types)
            assert "triangle" in types, (tag, types)


def _wedge_faces(b):
    return [
        np.array([b + 0, b + 2, b + 1]),
        np.array([b + 3, b + 4, b + 5]),
        np.array([b + 0, b + 1, b + 4, b + 3]),
        np.array([b + 1, b + 2, b + 5, b + 4]),
        np.array([b + 2, b + 0, b + 3, b + 5]),
    ]


def _tet_faces(b):
    return [
        np.array([b + 0, b + 2, b + 1]),
        np.array([b + 0, b + 1, b + 3]),
        np.array([b + 1, b + 2, b + 3]),
        np.array([b + 2, b + 0, b + 3]),
    ]


def test_polyhedra_with_alternating_node_counts_keep_their_cell_data():
    # A polyhedral run that mixes node counts is bucketed into one block per count, so a
    # bucket's cells are not contiguous in the file. cell_data has to follow the cells.
    pts = np.random.default_rng(0).random((16, 3))
    mesh = meshioplusplus.Mesh(
        pts,
        [
            ("polyhedron6", [_wedge_faces(0)]),
            ("polyhedron4", [_tet_faces(6)]),
            ("polyhedron6", [_wedge_faces(10)]),
        ],
        cell_data={"tag": [np.array([1.0]), np.array([2.0]), np.array([3.0])]},
    )

    for writer, reader, label in (
        (_vtu.write, _vtu.read, "python"),
        (meshioplusplus.vtu.write, meshioplusplus.vtu.read, "shim"),
    ):
        with tempfile.TemporaryDirectory() as d:
            p = pathlib.Path(d) / "alt.vtu"
            writer(p, mesh)
            out = reader(p)
        by_type = {}
        for cell, tag in zip(out.cells, out.cell_data["tag"]):
            by_type[cell.type] = np.asarray(tag).tolist()
        assert by_type == {"polyhedron6": [1.0, 3.0], "polyhedron4": [2.0]}, (
            label,
            by_type,
        )


# tools/gen_feconv_quirk_fixtures.py: one pyramid on one polyhedral cube, in
# every framing FEconv's samples use (and base64 appended, which they do not).
_QUIRKS = pathlib.Path(__file__).resolve().parent / "meshes" / "vtu"


def _core_vtu_read(path):
    from meshioplusplus import _core

    return _core.vtu_read(str(path))


@pytest.mark.parametrize("reader", [_core_vtu_read, _vtu.read], ids=["core", "python"])
@pytest.mark.parametrize(
    "name", ["raw_bigendian", "raw_zlib", "base64_appended", "binary_bigendian"]
)
def test_appended_and_bigendian_framings(reader, name):
    from meshioplusplus import _core

    if (
        reader is _core_vtu_read
        and name == "raw_zlib"
        and not getattr(_core, "__has_zlib__", False)
    ):
        pytest.skip("this build has no zlib")
    mesh = reader(_QUIRKS / f"{name}.vtu")
    assert [c.type for c in mesh.cells] == ["pyramid", "polyhedron8"]
    np.testing.assert_array_equal(mesh.cells[0].data, [[4, 5, 6, 7, 8]])
    faces = [np.asarray(f).tolist() for f in mesh.cells[1].data[0]]
    assert faces[0] == [3, 0, 4, 7] and faces[-1] == [4, 5, 6, 7]
    np.testing.assert_array_equal(mesh.points[8], [0.5, 0.5, 2.0])
    np.testing.assert_array_equal(mesh.point_data["heat"], np.arange(9.0))
    assert [np.asarray(d).tolist() for d in mesh.cell_data["material"]] == [[7], [9]]
    assert mesh.cell_data["material"][0].dtype.isnative


@pytest.mark.parametrize("reader", [_core_vtu_read, _vtu.read], ids=["core", "python"])
def test_two_pieces_with_polyhedra_merge(reader):
    mesh = reader(_QUIRKS / "two_pieces_polyhedron.vtu")
    assert len(mesh.points) == 18
    assert [c.type for c in mesh.cells] == ["pyramid", "polyhedron8"] * 2
    # The second piece's node ids follow the first piece's nine points.
    np.testing.assert_array_equal(mesh.cells[2].data, [[13, 14, 15, 16, 17]])
    faces = [np.asarray(f).tolist() for f in mesh.cells[3].data[0]]
    assert faces[0] == [12, 9, 13, 16]
    # Cell data in both pieces survives; point data in only one is dropped.
    assert [np.asarray(d).tolist() for d in mesh.cell_data["material"]] == [
        [7],
        [9],
    ] * 2
    assert "heat" not in (mesh.point_data or {})


def test_two_pieces_metadata_matches_the_read():
    from meshioplusplus import _core

    meta = _core.read_metadata(str(_QUIRKS / "two_pieces_polyhedron.vtu"), "vtu")
    assert meta["num_points"] == 18
    assert meta["cell_data_names"] == ["material"]
    assert meta["point_data_names"] == []


def test_faceoffsets_after_an_ordinary_block_end_at_the_stream(tmp_path):
    # faceoffsets are END offsets into `faces`. After a non-polyhedral block
    # (-1 entries) the next polyhedron's offset used to be rebased on that -1,
    # one short, and VTK (9.4+, ParaView 6) refused the whole file.
    import re

    pts = np.random.default_rng(1).random((20, 3))
    mesh = meshioplusplus.Mesh(
        pts,
        [
            ("tetra", [[0, 1, 2, 3]]),
            ("polyhedron4", [_tet_faces(4)]),
            ("hexahedron", [list(range(8, 16))]),
            ("polyhedron4", [_tet_faces(16)]),
        ],
    )
    p = tmp_path / "mixed.vtu"
    _vtu.write(p, mesh, binary=False)
    text = p.read_text()

    def array(name):
        m = re.search(rf'Name="{name}"[^>]*>(.*?)</DataArray>', text, re.S)
        return [int(v) for v in m.group(1).split()]

    faces, faceoffsets = array("faces"), array("faceoffsets")
    assert faceoffsets[0] == -1 and faceoffsets[2] == -1
    assert faceoffsets[3] == len(faces)
    assert faceoffsets[1] == 1 + 4 * 4  # one tetrahedron: count + 4 x (3 + 1)
    out = _vtu.read(p)
    assert [c.type for c in out.cells] == [c.type for c in mesh.cells]

File: tests/cpp/test_vtu.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//

// External includes
#include <gtest/gtest.h>

#include <algorithm>
#include <chrono>
#include <map>
#include <fstream>
#include <iterator>
#include <limits>

// Project includes
#include "mesh_fixtures.hpp"
#include "meshioplusplus/formats/vtu.hpp"
#include "meshioplusplus/operations/stats.hpp"

namespace {
void rt(const mt::Mesh& mesh, bool binary, bool zlib) {
    mt::roundtrip([=](const std::string& p,
                      const mt::Mesh& m) { meshioplusplus::write_vtu(p, m, binary, zlib); },
                  [](const std::string& p) { return meshioplusplus::read_vtu(p); }, mesh, ".vtu");
}
}  // namespace

TEST(Vtu, AsciiTri) {
    rt(mt::tri_mesh(), false, false);
}
TEST(Vtu, AsciiTetHexQuad) {
    rt(mt::tet_mesh(), false, false);
    rt(mt::hex_mesh(), false, false);
    rt(mt::quad_mesh(), false, false);
}
TEST(Vtu, AsciiHybrid) {
    rt(mt::tri_quad_mesh(), false, false);
}
TEST(Vtu, BinaryUncompressed) {
    rt(mt::tri_mesh(), true, false);
    rt(mt::tet_mesh(), true, false);
}
#ifdef MESHIOPLUSPLUS_HAS_ZLIB
TEST(Vtu, BinaryZlib) {
    rt(mt::tri_mesh(), true, true);
    rt(mt::hex_mesh(), true, true);
}
#endif

// --- VTK_POLYHEDRON (type 42) ------------------------------------------------

TEST(Vtu, PolyhedronRoundTrip) {
    // Both directions were refused before v9.19.0: the writer rejected any
    // polyhedron block, and the reader threw on merely SEEING the `faces`
    // array name -- before checking whether any cell was actually type 42.
    meshioplusplus::Mesh m;
    m.AssignPoints(mt::points_from(
        {{0, 0, 0}, {1, 0, 0}, {1, 1, 0}, {0, 1, 0}, {0, 0, 1}, {1, 0, 1}, {1, 1, 1}, {0, 1, 1}}));
    m.AddPolyhedronBlock(
        "polyhedron8",
        {{{0, 3, 2, 1}, {4, 5, 6, 7}, {0, 1, 5, 4}, {2, 3, 7, 6}, {0, 4, 7, 3}, {1, 2, 6, 5}}});

    for (bool binary : {false, true}) {
        const std::string p = mt::temp_path(binary ? "_poly_b.vtu" : "_poly_a.vtu");
        meshioplusplus::write_vtu(p, m, binary, /*zlib=*/false);
        const meshioplusplus::Mesh back = meshioplusplus::read_vtu(p);
        ASSERT_EQ(back.NumCellBlocks(), 1u) << "binary=" << binary;
        const auto cb = back.Cells(0);
        EXPECT_TRUE(cb.IsPolyhedron());
        EXPECT_EQ(cb.NumFaces(0), 6u);
        EXPECT_NEAR(meshioplusplus::compute_stats(back).mUnsignedVolume, 1.0, 1e-12);
        std::error_code ec;
        std::filesystem::remove(p, ec);
    }
}

TEST(Vtu, PolyhedraMixWithOtherCellTypes) {
    // The Python reference forbids mixing (_vtu.py's "cannot mix polyhedral
    // cells with other cell types"), but the FORMAT does not: `faceoffsets`
    // carries -1 for a non-polyhedral cell, which IS the mixing mechanism. An
    // OpenFOAM mesh always mixes, so inheriting that restriction would defeat
    // the point.
    meshioplusplus::Mesh m;
    m.AssignPoints(mt::points_from({{0, 0, 0},
                                    {1, 0, 0},
                                    {1, 1, 0},
                                    {0, 1, 0},
                                    {0, 0, 1},
                                    {1, 0, 1},
                                    {1, 1, 1},
                                    {0, 1, 1},
                                    {2, 0, 0},
                                    {2, 1, 0},
                                    {2, 1, 1},
                                    {2, 0, 1}}));
    m.AddCellBlock("hexahedron", mt::conn_from({{0, 1, 2, 3, 4, 5, 6, 7}}));
    m.AddPolyhedronBlock("polyhedron8", {{{1, 2, 10, 5},
                                          {8, 11, 10, 9},
                                          {1, 8, 9, 2},
                                          {5, 10, 11, 4 + 4},
                                          {1, 5, 11, 8},
                                          {2, 9, 10, 2}}});

    const std::string p = mt::temp_path("_poly_mixed.vtu");
    meshioplusplus::write_vtu(p, m, /*binary=*/false, /*zlib=*/false);
    // Both arrays must be present, and faceoffsets must carry the -1 sentinel
    // for the hexahedron -- that is the whole mixing contract.
    std::ifstream in(p, std::ios::binary);
    ASSERT_TRUE(in.good());
    std::string text((std::istreambuf_iterator<char>(in)), std::istreambuf_iterator<char>());
    EXPECT_NE(text.find("\"faces\""), std::string::npos);
    EXPECT_NE(text.find("\"faceoffsets\""), std::string::npos);
    EXPECT_NE(text.find("-1"), std::string::npos);

    const meshioplusplus::Mesh back = meshioplusplus::read_vtu(p);
    bool saw_hex = false, saw_poly = false;
    for (const auto cb : back.CellRange()) {
        if (cb.IsPolyhedron())
            saw_poly = true;
        else if (std::string(cb.Type()) == "hexahedron")
            saw_hex = true;
    }
    EXPECT_TRUE(saw_hex) << "the hexahedron was lost in a mixed polyhedral file";
    EXPECT_TRUE(saw_poly) << "the polyhedron was lost in a mixed file";
    std::error_code ec;
    std::filesystem::remove(p, ec);
}

TEST(Vtu, ThreePolyhedraWithAnInterleavedCellReadCorrectly) {
    // Regression test for the O(ncells^2) `faceoffsets` rescan in
    // reconstruct_cells: with only one polyhedron (PolyhedronRoundTrip) or one
    // polyhedron preceded by one other cell (PolyhedraMixWithOtherCellTypes),
    // the inner rescan never actually runs more than once, so neither test
    // above can distinguish the correct fix (a running "last end offset"
    // carried across the WHOLE loop) from two plausible wrong ones: resetting
    // that running value at the start of each outer-loop run, or at the start
    // of each polyhedral run specifically. Both wrong forms are only exposed
    // when a polyhedral RUN is preceded by another polyhedral run separated by
    // an intervening non-polyhedral one -- exactly this mesh's block order:
    // polyhedron, polyhedron, hexahedron, polyhedron (four distinct blocks,
    // written and therefore read back as three vtk-type runs: [42,42], [12],
    // [42]). If the third polyhedron's face-stream start were computed wrong,
    // its decoded cell would come out non-manifold (wrong node count / a
    // volume of zero or a nonsensical value) rather than throwing, so the
    // volume check below is the one that actually catches it.
    auto unit_cube_faces = [](std::int64_t base) {
        return std::vector<std::vector<std::int64_t>>{
            {base + 0, base + 3, base + 2, base + 1}, {base + 4, base + 5, base + 6, base + 7},
            {base + 0, base + 1, base + 5, base + 4}, {base + 2, base + 3, base + 7, base + 6},
            {base + 0, base + 4, base + 7, base + 3}, {base + 1, base + 2, base + 6, base + 5}};
    };

    std::vector<std::vector<double>> pts;
    for (int block = 0; block < 4; ++block) {
        const double ox = 2.0 * block;  // disjoint, non-overlapping cubes
        for (const auto& c : {std::vector<double>{0, 0, 0},
                              {1, 0, 0},
                              {1, 1, 0},
                              {0, 1, 0},
                              {0, 0, 1},
                              {1, 0, 1},
                              {1, 1, 1},
                              {0, 1, 1}})
            pts.push_back({c[0] + ox, c[1], c[2]});
    }

    meshioplusplus::Mesh m;
    m.AssignPoints(mt::points_from(pts));
    m.AddPolyhedronBlock("polyhedron8", {unit_cube_faces(0)});
    m.AddPolyhedronBlock("polyhedron8", {unit_cube_faces(8)});
    m.AddCellBlock("hexahedron", mt::conn_from({{16, 17, 18, 19, 20, 21, 22, 23}}));
    m.AddPolyhedronBlock("polyhedron8", {unit_cube_faces(24)});

    const std::string p = mt::temp_path("_poly_interleaved.vtu");
    meshioplusplus::write_vtu(p, m, /*binary=*/false, /*zlib=*/false);
    const meshioplusplus::Mesh back = meshioplusplus::read_vtu(p);
    std::error_code ec;
    std::filesystem::remove(p, ec);

    // Two polyhedron<8> blocks (the two runs get bucketed together by unique
    // node count, since bucketing is per-run) plus one hexahedron block.
    std::size_t num_poly_cells = 0, num_hex_cells = 0;
    for (const auto cb : back.CellRange()) {
        if (cb.IsPolyhedron()) {
            for (std::size_t r = 0; r < cb.NumCells(); ++r) {
                EXPECT_EQ(cb.NumFaces(r), 6u) << "cell " << num_poly_cells;
                ++num_poly_cells;
            }
        } else if (std::string(cb.Type()) == "hexahedron") {
            num_hex_cells += cb.NumCells();
        }
    }
    EXPECT_EQ(num_poly_cells, 3u);
    EXPECT_EQ(num_hex_cells, 1u);

    // Every cell -- all four blocks -- must be a unit cube; a wrong
    // begin-offset for the third polyhedron would slice into an unrelated
    // part of the faces stream (or throw as out-of-range), never quietly
    // reproduce the right geometry.
    const auto stats = meshioplusplus::compute_stats(back);
    EXPECT_NEAR(stats.mUnsignedVolume, 4.0, 1e-12);
}

namespace {

/// Seconds to read back an ASCII VTU of @p NumCells identical unit-cube
/// polyhedra (best of two reads, which damps scheduler jitter). Every cell
/// reuses the same 8 points / 6 faces on purpose: construction cost and memory
/// are irrelevant here, only the read-side cell-reconstruction loop is timed.
double vtu_polyhedron_read_seconds(std::size_t NumCells) {
    meshioplusplus::Mesh m;
    m.AssignPoints(mt::points_from(
        {{0, 0, 0}, {1, 0, 0}, {1, 1, 0}, {0, 1, 0}, {0, 0, 1}, {1, 0, 1}, {1, 1, 1}, {0, 1, 1}}));
    std::vector<std::vector<std::vector<std::int64_t>>> cells(
        NumCells,
        std::vector<std::vector<std::int64_t>>{
            {0, 3, 2, 1}, {4, 5, 6, 7}, {0, 1, 5, 4}, {2, 3, 7, 6}, {0, 4, 7, 3}, {1, 2, 6, 5}});
    m.AddPolyhedronBlock("polyhedron8", std::move(cells));

    const std::string p = mt::temp_path("_poly_scaling.vtu");
    meshioplusplus::write_vtu(p, m, /*binary=*/false, /*zlib=*/false);

    double best = std::numeric_limits<double>::infinity();
    for (int rep = 0; rep < 2; ++rep) {
        const auto t0 = std::chrono::steady_clock::now();
        const meshioplusplus::Mesh back = meshioplusplus::read_vtu(p);
        const double elapsed =
            std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();
        EXPECT_EQ(back.NumCellBlocks(), 1u);
        EXPECT_EQ(back.Cells(0).NumCells(), NumCells);
        best = std::min(best, elapsed);
    }

    std::error_code ec;
    std::filesystem::remove(p, ec);
    return best;
}

}  // namespace

TEST(Vtu, PolyhedraWithAlternatingNodeCountsKeepTheirCellData) {
    // A polyhedral run mixing node counts is bucketed into one block per count, so a
    // bucket's cells are not contiguous in the file: wedge(1), tet(2), wedge(3) must read
    // back as polyhedron6 -> {1, 3} and polyhedron4 -> {2}, not the file-order slices
    // {1, 2} and {3} that bucket-by-bucket slicing used to hand out.
    auto wedge = [](std::int64_t b) {
        return std::vector<std::vector<std::int64_t>>{{b, b + 2, b + 1},
                                                      {b + 3, b + 4, b + 5},
                                                      {b, b + 1, b + 4, b + 3},
                                                      {b + 1, b + 2, b + 5, b + 4},
                                                      {b + 2, b, b + 3, b + 5}};
    };
    auto tet = [](std::int64_t b) {
        return std::vector<std::vector<std::int64_t>>{
            {b, b + 2, b + 1}, {b, b + 1, b + 3}, {b + 1, b + 2, b + 3}, {b + 2, b, b + 3}};
    };
    std::vector<std::vector<double>> pts;
    for (int i = 0; i < 16; ++i)
        pts.push_back({0.1 * i, (i % 3) * 0.5, (i % 5) * 0.25});

    meshioplusplus::Mesh m;
    m.AssignPoints(mt::points_from(pts));
    m.AddPolyhedronBlock("polyhedron6", {wedge(0)});
    m.AddPolyhedronBlock("polyhedron4", {tet(6)});
    m.AddPolyhedronBlock("polyhedron6", {wedge(10)});
    std::vector<meshioplusplus::NDArray> tags;
    for (double v : {1.0, 2.0, 3.0}) {
        meshioplusplus::NDArray blk(meshioplusplus::DType::Float64, {std::size_t{1}});
        blk.As<double>()[0] = v;
        tags.push_back(std::move(blk));
    }
    m.AddCellData("tag", std::move(tags));

    const std::string p = mt::temp_path("_poly_alternating.vtu");
    meshioplusplus::write_vtu(p, m, /*binary=*/false, /*zlib=*/false);
    const meshioplusplus::Mesh back = meshioplusplus::read_vtu(p);
    std::error_code ec;
    std::filesystem::remove(p, ec);

    ASSERT_EQ(back.NumCellBlocks(), 2u);
    ASSERT_TRUE(back.HasCellData("tag"));
    std::map<std::string, std::vector<double>> tags_by_type;
    for (std::size_t b = 0; b < back.NumCellBlocks(); ++b) {
        const meshioplusplus::NDArray& d = back.CellData("tag", b);
        ASSERT_EQ(d.Size(), back.Cells(b).NumCells());
        tags_by_type[std::string(back.Cells(b).Type())] =
            std::vector<double>(d.As<double>(), d.As<double>() + d.Size());
    }
    EXPECT_EQ(tags_by_type["polyhedron6"], (std::vector<double>{1.0, 3.0}));
    EXPECT_EQ(tags_by_type["polyhedron4"], (std::vector<double>{2.0}));
}

TEST(Vtu, PolyhedronReconstructionIsNotQuadraticInCellCount) {
    // Direct regression test for the roadmap's own probe: the pre-fix
    // reconstruct_cells rescanned `faceoffsets` from index 0 for EVERY cell, so
    // total work was O(ncells^2) regardless of how much face data there was. The
    // fix makes it O(ncells + |faces|).
    //
    // This asserts how the time SCALES, not how long it takes. It used to assert
    // an absolute 3 s budget for 100k cells, which is a statement about the
    // machine as much as the code: under the instrumented coverage build the
    // linear read already used ~90% of it (4.7 s of ctest wall time on master),
    // so the test failed intermittently on unrelated changes. Growing the cell
    // count 5x costs about 5x when reconstruction is linear and about 25x when
    // it is quadratic, whatever the machine or instrumentation; the threshold
    // sits between the two (their geometric mean is ~11).
    constexpr std::size_t kSmall = 20000;
    constexpr std::size_t kLarge = 100000;
    constexpr double kMaxRatio = 12.0;

    const double t_small = vtu_polyhedron_read_seconds(kSmall);
    const double t_large = vtu_polyhedron_read_seconds(kLarge);
    // A floor keeps the ratio finite if a fast machine resolves the small read as ~0.
    const double ratio = t_large / std::max(t_small, 1e-3);

    GTEST_LOG_(INFO) << "read " << kSmall << " cells in " << t_small << "s, " << kLarge << " in "
                     << t_large << "s: x" << ratio << " for x" << kLarge / kSmall << " the cells";
    EXPECT_LT(ratio, kMaxRatio) << "reading " << kLarge / kSmall << "x the cells took " << ratio
                                << "x as long (" << t_small << "s -> " << t_large
                                << "s); linear is ~" << kLarge / kSmall
                                << "x, so reconstruct_cells looks quadratic again";
}
